pkg/controller/controller_test.go

```diff
@@ -2781,3 +2747,190 @@ func TestProvisionFromPVC(t *testing.T) {
 		}
 	}
 }
+
+func TestProvisionWithMigration(t *testing.T) {
+	var requestBytes int64 = 100000
+	var inTreePluginName = "in-tree-plugin"
+
+	deletePolicy := v1.PersistentVolumeReclaimDelete
+	testcases := []struct {
+		name              string
+		provisioner       string
+		annotation        map[string]string
+		expectTranslation bool
+		expectCreate      bool
+		expectErr         bool
+	}{
+		{
+			name:              "provision with migration on",
+			provisioner:       inTreePluginName,
+			annotation:        map[string]string{annStorageProvisioner: driverName},
+			expectCreate:      true,
+			expectTranslation: true,
+		},
+		{
+			name:              "provision without migration for native CSI",
+			provisioner:       driverName,
+			annotation:        map[string]string{annStorageProvisioner: driverName},
+			expectCreate:      true,
+			expectTranslation: false,
+		},
+		{
+			name:        "ignore in-tree PVC when provisioned by in-tree",
+			provisioner: inTreePluginName,
+			annotation:  map[string]string{annStorageProvisioner: inTreePluginName},
+			expectErr:   true,
+		},
+	}
+
+	for k, tc := range testcases {
+		t.Logf("running test: %s", tc.name)
+		// Set up test
+		tmpdir := tempDir(t)
+		defer os.RemoveAll(tmpdir)
+		mockController, driver, _, controllerServer, csiConn, err := createMockServer(t, tmpdir)
+		if err != nil {
+			t.Fatal(err)
+		}
+		mockTranslator := NewMockProvisionerCSITranslator(mockController)
+		defer mockController.Finish()
```

Comment:
`defer` executes when the enclosing function returns, and not at the end of each loop iteration. Is that intended and desirable? (Mainly asking because I don't know how this mock API works.)

Reply:
no I think it should be in the end of the loop iteration - nested it inside `t.Run` so that should work now